src/main/java/ee/carlrobert/codegpt/codecompletions/InlineCodeCompletionRenderer.java

```diff
@@ -0,0 +1,241 @@
+  }
+
+  @Override
+  public void beforeDocumentChange(@NotNull DocumentEvent event) {
+    resetSuggestion();
+  }
+
+  void restartTimer() {
+    typingTimer.restart();
+  }
+
+  private void resetSuggestion() {
+    restartTimer();
+    ActionManager actionManager = ActionManager.getInstance();
+    actionManager.unregisterAction(ACTION_ID);
+    removeInlay();
+  }
+
+  private void enableSuggestions() {
+    if (!typingTimer.isRunning()) {
+      typingTimer.start();
+    }
+  }
+
+  void disableSuggestions() {
+    resetSuggestion();
+    typingTimer.stop();
+  }
+
+  @Override
+  public void keyTyped(KeyEvent e) {
+    resetSuggestion();
+  }
+
+  @Override
+  public void keyPressed(KeyEvent e) {
+    resetSuggestion();
+  }
+
+  @Override
+  public void keyReleased(KeyEvent e) {
+  }
+
+  @Override
+  public void caretPositionChanged(@NotNull CaretEvent event) {
+    resetSuggestion();
+  }
+
+  @Override
+  public void selectionChanged(@NotNull SelectionEvent e) {
+    TextRange range = e.getNewRange();
+    int rangeLength = range.getLength();
+    if (rangeLength != 0) {
+      disableSuggestions();
+    } else {
+      enableSuggestions();
+    }
+  }
+
+  @Override
+  public int calcWidthInPixels(@NotNull Inlay inlay) {
+    FontMetrics fontMetrics = editor.getContentComponent()
+        .getFontMetrics(editor.getColorsScheme().getFont(EditorFontType.PLAIN));
+    return fontMetrics.stringWidth(inlayText);
+  }
+
+
+  @Override
+  public void paint(@NotNull Inlay inlay, @NotNull Graphics2D g, @NotNull Rectangle2D targetRegion,
+      @NotNull TextAttributes textAttributes) {
+    // Create custom TextAttributes for different font type or color
+    TextAttributes customAttributes = new TextAttributes();
+    customAttributes.setFontType(Font.ITALIC);
+    customAttributes.setForegroundColor(JBColor.GRAY);
+
+    // Apply the custom attributes to the Graphics object
+    Font font = editor.getColorsScheme().getFont(EditorFontType.ITALIC)
+        .deriveFont(customAttributes.getFontType());
+    g.setFont(font);
+    g.setColor(customAttributes.getForegroundColor());
+
+    // Calculate the ascent to render the text on the same line as the caret
+    int ascent = editor.getAscent();
+
+    // Draw the string with the custom attributes
+    g.drawString(inlayText, (int) targetRegion.getX(), (int) targetRegion.getY() + ascent);
+  }
+
+  public String getInlayText() {
+    return inlayText;
+  }
+
+  private static Message createFimMessage(int offset, Document document) {
+    int begin = Integer.max(0, offset - MAX_OFFSET);
+    int end = Integer.min(document.getTextLength(), offset + MAX_OFFSET);
+    var before = document.getText(new TextRange(begin, offset));
+    var after = document.getText(new TextRange(offset, end));
+    if (SettingsState.getInstance().getSelectedService() == ServiceType.LLAMA_CPP) {
+      // Use Messages prompt as input_prefix and response field as input_suffix
+      return new Message(before, after);
```

Comment:
Probably this should be done cleaner, but works to just try the `/infill`, since its a special case limited to LLAMA

Reply:
I would refactor this by introducing a `LLamaInfillMessage` which extends `Message` and adds `inputPrefix` and `inputSuffix`, so this get cleaner

Reply:
I've refactored this by introducing a new class `InfillRequestDetails` that contains the `prefix` and `suffix` fields, which are passed onto the `CodeCompletionRequestProvider` which handles the actual infill request building.
In future, I want to separate code completion from chat completion to achieve some kind of modularity between them.